src/Compilers/CSharp/Test/Semantic/Semantics/UsingDeclarationTests.cs

```diff
@@ -0,0 +1,396 @@
+                //         using var x = (IDisposable)null;
+                Diagnostic(ErrorCode.WRN_UnreachableCode, "using").WithLocation(10, 9)
+                );
+        }
+
+        [Fact]
+        public void DisallowGoToForwardAcrossMultipleUsingDeclarationsGivesOnlyOneDiagnostic()
+        {
+            var source = @"
+using System;
+class C
+{
+    static void Main()
+    {
+        goto label1;
+        using var x = (IDisposable)null;
+        using var y = (IDisposable)null;
+
+        label1:
+        return;
+    }
+}
+";
+            CreateCompilation(source).VerifyDiagnostics(
+                // (7,9): error CS8641: A goto target within the same block can not cross a using declaration.
+                //         goto label1;
+                Diagnostic(ErrorCode.ERR_GoToForwardJumpOverUsingVar, "goto label1;").WithLocation(7, 9),
+                // (8,9): warning CS0162: Unreachable code detected
+                //         using var x = (IDisposable)null;
+                Diagnostic(ErrorCode.WRN_UnreachableCode, "using").WithLocation(8, 9)
+                );
+        }
+
+        [Fact]
+        public void DisallowMultipleGoToForwardAcrossMultipleUsingDeclarations()
+        {
+            var source = @"
+using System;
+class C
+{
+    static void Main()
+    {
+        goto label1;
+        using var x = (IDisposable)null;
+        goto label2;
+        using var y = (IDisposable)null;
+
+        label1:
+        label2:
+        return;
+    }
+}
+";
+            CreateCompilation(source).VerifyDiagnostics(
+                // (7,9): error CS8641: A goto target within the same block can not cross a using declaration.
+                //         goto label1;
+                Diagnostic(ErrorCode.ERR_GoToForwardJumpOverUsingVar, "goto label1;").WithLocation(7, 9),
+                // (8,9): warning CS0162: Unreachable code detected
+                //         using var x = (IDisposable)null;
+                Diagnostic(ErrorCode.WRN_UnreachableCode, "using").WithLocation(8, 9),
+                // (9,9): error CS8641: A goto target can not be after any using declarations.
+                //         goto label2;
+                Diagnostic(ErrorCode.ERR_GoToForwardJumpOverUsingVar, "goto label2;").WithLocation(9, 9)
+                );
+        }
+
+        [Fact]
+        public void DisallowGoToBackwardsAcrossUsingDeclarations()
+        {
+            var source = @"
+using System;
+class C
+{
+    static void Main()
+    {
+        label1:
+        using var x = (IDisposable)null;
+
+        goto label1;
+    }
+}
+";
+            CreateCompilation(source).VerifyDiagnostics(
+                // (10,9): error CS8641: A goto target within the same block can not cross a using declaration.
+                //         goto label1;
+                Diagnostic(ErrorCode.ERR_GoToBackwardJumpOverUsingVar, "goto label1;").WithLocation(10, 9)
+                );
+        }
```

Comment:
>} [](start = 8, length = 1)

Consider testing `goto case`:
```c#
switch (value)
{
    case A:
        using (GetDisposable());
        goto case A;
}
```

Reply:
That isn't valid, as we don't allow using declarations in raw switch blocks. We have a test for this already in UsingStatementTests.UsingVarInSwitchCase 

I'll move it over to the declarations class to make it clearer though.

Reply:
In that case, consider testing:
```C#
switch (value)
{
case A:
    {
        using (GetDisposable());
        goto case A;
    }
}
```

---
In reply to: [242682342](https://github.com/dotnet/roslyn/pull/31674#discussion_r242682342) [](ancestors = 242682342)